### PR TITLE
Split logical rule for eval from for train

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

We might want different sharding logical rule for eval compared with train, since the global train batch size and eval batch size are different. This PR enables this feature by introducing two new flags:
* `custom_mesh_and_rule_for_eval`
* `logical_axis_rules_for_eval`

When `custom_mesh_and_rule_for_eval` is specified (non-default value), an alternative logical rule is used for the eval pipeline. A typical use case would be using fractional batch size for eval, so that CP/TP like shardings are enabled (only for eval). Same axis is used as other shardings, e.g. FSDP/EP, for the train step.

Currently, we still use the same mesh for both train and eval steps. A potential future optimization would be split the meshes for train and eval, which is in theory possible  since train and eval functions are in different jit.It may required careful checks.

This PR only supports auto sharding. For explicit sharding, the logical rule split is way more complexed. I would defer the explicit sharding support to 2nd PR.

FIXES: b/535639231

# Tests

## Example 1: base rule for train, custom rule for eval
Run on v5p-8: we use FSDP+EP for train step, and EP-as-CP custom rule for eval,

```
smoke_train model_name=deepseek3-test per_device_batch_size=1 eval_per_device_batch_size=0.25 eval_interval=3 eval_steps=1 ici_expert_parallelism=4 custom_mesh_and_rule_for_eval=ep-as-cp debug_sharding=false profiler=xplane use_ring_of_experts=true  max_target_length=4096
```

[log](https://paste.googleplex.com/4710615229792256)

[xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-6166349024609357677)

## Example 2: custom rule for train, another custom rule for eval
Run on v5p-8: we use EP-as-DP for train step, and EP-as-CP custom rule for eval,

```
smoke_train model_name=deepseek3-test per_device_batch_size=1 eval_per_device_batch_size=0.25 eval_interval=3 eval_steps=1 ici_expert_parallelism=4 custom_mesh_and_rule_for_eval=ep-as-cp debug_sharding=false profiler=xplane use_ring_of_experts=true  max_target_length=4096  custom_mesh_and_rule=ep-as-dp sharding_tolerance=0.5 override_model_config=true base_num_decoder_layers=7
```

[log](https://paste.googleplex.com/5560138252550144)

[xprof](http://xprof.corp.google.com/trace_viewer/chengnuojin-14291228915244752956)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/utils/sharding.py [hunk 1]
```diff
@@ -51,6 +51,22 @@ def get_input_data_sharding(config, mesh):
...
+
+def get_logical_axis_rules_from_config(config):
+  """Get the logical axis rules from the config.
+  When we use pipeline parallelism or in eval step, we may use
+  a different logical axis rule than the one in config.
+  Plan to deprecate (b/536927795) and use get_logical_axis_rules instead.
+  """
+  if config.using_pipeline_parallelism or config.eval_interval > 1:
+    return None
+  return config.logical_axis_rules
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/utils/sharding.py:67]: I definitly agree that `eval_interval > 1` is not the best way for selecting logical rules. I have tried using flax meta data to collect logical rules but it gets some unit test failures. I will be working on b/536927795 to deprecate this usage as a followup step.

File: src/maxtext/models/qwen3.py [hunk 2]
```diff
@@ -612,7 +612,7 @@ def __call__(
     # mixed_qkvz: (B, S, H_k, 2*D_k + 2*D_v*V_per_K)
     mixed_qkvz = qkvz.reshape(new_shape_qkvz)
     if self.mesh is not None:
-      logical_rules = None if self.config.using_pipeline_parallelism else self.config.logical_axis_rules
+      logical_rules = get_logical_axis_rules()
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/models/qwen3.py:615]: Good point. Updated

File: src/maxtext/trainers/pre_train/train.py [hunk 3]
```diff
@@ -727,7 +728,7 @@ def training_loop_iteration(
       eval_batch = jax.device_put(eval_batch, sharding.get_input_data_sharding(config, mesh))
       if 0 < eval_steps <= eval_step_count:
         break
-      with jax.set_mesh(mesh), nn_partitioning.axis_rules(logical_axis_rules):
+      with jax.set_mesh(mesh), nn_partitioning.axis_rules(logical_axis_rules_for_eval):
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/trainers/pre_train/train.py:733]: Good point. Updated

File: src/maxtext/configs/base.yml [hunk 4]
```diff
@@ -616,6 +616,9 @@ input_data_sharding_logical_axes: ['activation_embed_and_logits_batch', 'activat
 # Determines which physical axis plays the role of context parallelism for input data processing and load balancing
 # only supports "context" or "expert" (when custom_mesh_and_rule=ep-as-cp)
 context_sharding: "context"
+# Set as true a different logical rule is used for eval step
+custom_mesh_and_rule_for_eval: ""
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/configs/base.yml]: updated

File: src/maxtext/configs/types.py [hunk 5]
```diff
@@ -2697,31 +2704,46 @@ def validate_num_moe_emb_chunks(self):
...
+          setattr(self, field, mesh_config[field])
+
+    # Handle eval custom mesh and rule
+    if self.custom_mesh_and_rule_for_eval is CustomRule.DEFAULT:
+      # Fallback to primary rule if eval is DEFAULT
+      self.custom_mesh_and_rule_for_eval = self.custom_mesh_and_rule
+    else:
+      eval_config = self._load_mesh_config_from_yaml(self.custom_mesh_and_rule_for_eval.value)
+      if "logical_axis_rules" in eval_config:
+        self.logical_axis_rules_for_eval = eval_config["logical_axis_rules"]
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/configs/types.py]: Updated.

File: src/maxtext/trainers/pre_train/train.py [hunk 6]
```diff
@@ -663,7 +663,8 @@ def training_loop_iteration(
 
   # Unpack immutable_data
   config = immutable_data["config"]  # for helpers
-  logical_axis_rules = immutable_data["logical_axis_rules"]
+  logical_axis_rules_train = immutable_data["logical_axis_rules_train"]
+  logical_axis_rules_for_eval = immutable_data["logical_axis_rules_for_eval"]
```
Comment (RissyRan) [inline src/maxtext/trainers/pre_train/train.py:667]: Shall we align the name from `logical_axis_rules_for_eval` to `logical_axis_rules_eval`?

File: src/maxtext/trainers/pre_train/train.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/trainers/pre_train/train.py:733]: I feel like the Gemini PR review shows better quality now.

File: src/maxtext/utils/train_utils.py [hunk 7]
```diff
@@ -191,13 +191,14 @@ def jit_train_and_eval_step(
   if config.enable_diloco:
     train_step_partial = functools.partial(train_step, model, config, state_mesh_shardings, params_shardings)
     train_step = diloco.build_diloco_train_step(config, train_step_partial, mesh=mesh)
-  data_sharding = sharding.get_input_data_sharding(config, mesh)
+  train_data_sharding = sharding.get_input_data_sharding(config, mesh, rules=config.logical_axis_rules)
```
Comment (RissyRan) [inline src/maxtext/utils/train_utils.py]: I see some misalignment with `logical_axis_rules_train` and `logical_axis_rules`. Do you think we could let them aligned? i.e. we could use `logical_axis_rules` for training stage if that makes things simple?

File: src/maxtext/utils/sharding.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/utils/sharding.py:67]: Are you trying to distinguish if eval is turned on? If so, wondering if `eval_per_device_batch_size` a better indicator. I see default is 0: https://github.com/AI-Hypercomputer/maxtext/blob/2ce38667035bdc24d4cca402bcca1f8ef90f2b8c/src/maxtext/configs/base.yml#L696. 

In MLPerf 6.0 runs, we set eval_interval=1. Or maybe you could check if eval_interval as -1: https://github.com/AI-Hypercomputer/maxtext/blob/2ce38667035bdc24d4cca402bcca1f8ef90f2b8c/src/maxtext/configs/base.yml#L986

**Review by RissyRan — COMMENTED** [review]
Thanks for the change! A few minor comments for naming. 

I didn't see you test feature for `logical_axis_rules_for_eval`. Do you forsee potential conflict when applying separately?

File: src/maxtext/trainers/pre_train/train.py (hunk 6 above)
Comment (NuojCheng) [reply src/maxtext/trainers/pre_train/train.py:667]: I updated the names to
* `logical_axis_rules_for_train`
* `logical_axis_rules_for_eval`
for consistency. Good point!

File: src/maxtext/trainers/pre_train/train.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/trainers/pre_train/train.py:733]: Yes I agree!

File: src/maxtext/utils/sharding.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/utils/sharding.py:67]: Thank you for the suggestion! Updated using `eval_per_device_batch_size > 0` now.

File: src/maxtext/utils/train_utils.py (hunk 7 above)
Comment (NuojCheng) [reply src/maxtext/utils/train_utils.py]: Agree! Now they are
* `data_sharding_for_train` 
* `data_sharding_for_eval`

**Review by gobbleturk — APPROVED** [review]

File: src/maxtext/utils/sharding.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/maxtext/utils/sharding.py:67]: The logic in https://github.com/AI-Hypercomputer/maxtext/blob/bff4616de318c48c6e5fafbd65d7cf7b5c1bae07/src/maxtext/configs/types.py#L2811 sets eval pdb same with pdb. I will use `eval_interval != -1` instead.

**Review by RissyRan — APPROVED** [review]

**Comment by NuojCheng** [comment]
Regarding to question
> I didn't see you test feature for logical_axis_rules_for_eval. Do you forsee potential conflict when applying separately?

I added an integration test protecting this feature. Thank you for bringing it up!
